### ISSUE TITLE
VTK 9.4 Testing Failures

### ISSUE THREAD

**Issue by banesullivan** [pyvista/pyvista#6731]
### Describe what maintenance you would like added.

VTK 9.4 appears to be on the horizon with release candidates and we have begun testing PyVista against this version. As tradition there are a slew of testing failures to address!

This issue will be where we tract the progress of these failures from a meta level and we will create other issues/PRs referencing this meta issue as needed.

cc @pyvista/kitware 

### Core Testing failures

Seeing ~277 failures with the overwhelming majority (~256) coming from the `test_text3d_source_parameters` test and a check in our `copy_from` method which is showing up in other tests (and also why are there 256 parameterized runs of `test_text3d_source_parameters`?!?! 🤯 )

The remaining ~21 failures are listed in the following section (captured from `pytest -v tests/core -k 'not test_text3d_source_parameters'`

<details>

```
================================================== short test summary info ===================================================
FAILED tests/core/test_dataobject.py::test_pickle_serialize_deserialize[vtk] - AssertionError: assert <vtkmodules.v...x742809ec4580> == vtkStructured...able: (none)\n\n
FAILED tests/core/test_dataobject.py::test_pickle_serialize_deserialize[xml] - AssertionError: assert <vtkmodules.v...x742809ddf7c0> == vtkStructured...able: (none)\n\n
FAILED tests/core/test_dataobject.py::test_pickle_serialize_deserialize[legacy] - AssertionError: assert <vtkmodules.v...x742809ddfdc0> == vtkStructured...able: (none)\n\n
FAILED tests/core/test_composite.py::test_multi_block_eq - AssertionError: assert MultiBlock (0...0.000, 282.130 == MultiBlock (0...0.000, 282.130
FAILED tests/core/test_dataobject.py::test_pickle_multiblock[vtk] - AssertionError: assert MultiBlock (0...0.000, 282.130 == MultiBlock (0...0.000, 282.130
FAILED tests/core/test_dataobject.py::test_uniform_eq - AssertionError: assert ImageData (0x... Arrays:     2 == ImageData (0x... Arrays:     2
FAILED tests/core/test_geometric_sources.py::test_text3d_source_modified[kwarg_tuple5] - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_geometric_sources.py::test_text3d_source_empty_string - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_geometric_sources.py::test_text3d_source - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_geometric_sources.py::test_text3d_source_update - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_geometric_sources.py::test_text3d_source_output - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_geometric_objects.py::test_text_3d - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_geometric_sources.py::test_text3d_source_modified[kwarg_tuple0] - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_grid.py::test_create_image_data_from_specs - AssertionError: assert ImageData (0x... Arrays:     0 == ImageData (0x... Arrays:     0
FAILED tests/core/test_geometric_sources.py::test_text3d_source_modified[kwarg_tuple1] - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_geometric_sources.py::test_text3d_source_modified[kwarg_tuple2] - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_geometric_sources.py::test_text3d_source_modified[kwarg_tuple3] - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_geometric_sources.py::test_text3d_source_modified[kwarg_tuple4] - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/core/test_partitioned.py::test_slice_defaults - assert PartitionedDa...artitions    5 == PartitionedDa...artitions    5
FAILED tests/core/test_polydata.py::test_flip_normals - assert False
FAILED tests/core/test_reader.py::test_xmlpartitioneddatasetreader - AttributeError: 'PartitionedDataSet' object has no attribute '_association_bitarray_names'
ERROR tests/core/test_validation.py
================== 21 failed, 2442 passed, 18 skipped, 3 xfailed, 13 warnings, 1 error in 62.57s (0:01:02) ===================
```
</details>

### Plotting Testing failures

Let's resolve the core API failures before we dig too deep into these as core API failures are likely causing many of these plotting test failures.

Results output below `pytest -v tests/plotting` results in ~47 failures

<details>

```
================================================== short test summary info ===================================================
FAILED tests/plotting/test_charts.py::test_chart_2d - assert not True
FAILED tests/plotting/test_charts.py::test_axis_range - AssertionError: assert 0 == 'auto'
FAILED tests/plotting/test_charts.py::test_axis_scale - assert not True
FAILED tests/plotting/test_charts.py::test_axis_visible - assert not True
FAILED tests/plotting/test_charts.py::test_axis_tick_locations - TypeError: 'vtkmodules.vtkCommonCore.vtkStringArray' object is not iterable
FAILED tests/plotting/test_plotter.py::test_legend_font - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_plotting.py::test_add_box_axes - RuntimeError: test_add_box_axes Exceeded image regression error of 500.0 with an image error equal to: 1765.5738562091378
FAILED tests/plotting/test_plotting.py::test_add_legend - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_plotting.py::test_legend_circle_face - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_plotting.py::test_legend_rectangle_face - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_plotting.py::test_legend_subplots - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_plotting.py::test_add_axes_twice - RuntimeError: test_add_axes_twice Exceeded image regression error of 500.0 with an image error equal to: 2544.022222222311
FAILED tests/plotting/test_plotting.py::test_hide_axes - RuntimeError: test_hide_axes Exceeded image regression error of 500.0 with an image error equal to: 2608.3725490197035
FAILED tests/plotting/test_plotting.py::test_add_axes_parameters - RuntimeError: test_add_axes_parameters Exceeded image regression error of 500.0 with an image error equal to: 2487.058823...
FAILED tests/plotting/test_plotting.py::test_plot_composite_many_options - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_plotting.py::test_direction_objects[Text3D-pos] - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/plotting/test_plotting_utilities.py::test_plot_boundaries - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_plotting.py::test_direction_objects[Text3D-neg] - TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ...
FAILED tests/plotting/test_renderer.py::test_add_legend_loc[center] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_using_add_legend - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_legend_no_face - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_plotting_utilities.py::test_ray_trace_plot - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_bad_legend_origin_and_size - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_using_add_legend_with_glyph - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_remove_legend - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_legend_loc[upper right] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_face[-] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_legend_loc[upper left] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_using_add_legend_only_labels - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_face[^] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_legend_loc[lower left] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_face[o] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_none_face - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_legend_loc[lower right] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_face[r] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_add_entry_warning - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_legend_loc[center left] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_face[None] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_face[face5] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_legend_background_opacity - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_legend_loc[center right] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_legend_loc[lower center] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_from_glyph - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_add_legend_loc[upper center] - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_renderer.py::test_legend_from_multiple_glyph - TypeError: arguments do not match any overloaded methods
FAILED tests/plotting/test_theme.py::test_trame_config - assert not True
FAILED tests/plotting/test_utilities.py::test_start_xvfb - OSError: Please install Xvfb with:
ERROR tests/plotting/test_actor.py
ERROR tests/plotting/test_plotter.py::test_legend_font - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_add_box_axes - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_add_legend - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_legend_circle_face - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_legend_rectangle_face - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_legend_subplots - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_add_axes_twice - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_hide_axes - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_add_axes_parameters - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_plot_composite_many_options - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_direction_objects[Text3D-pos] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting_utilities.py::test_plot_boundaries - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting.py::test_direction_objects[Text3D-neg] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_loc[center] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_using_add_legend - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_no_face - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_plotting_utilities.py::test_ray_trace_plot - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_bad_legend_origin_and_size - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_remove_legend - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_using_add_legend_with_glyph - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_loc[upper right] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_face[-] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_loc[upper left] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_using_add_legend_only_labels - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_face[^] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_loc[lower left] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_face[o] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_none_face - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_loc[lower right] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_face[r] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_add_entry_warning - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_face[None] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_loc[center left] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_loc[center right] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_background_opacity - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_face[face5] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_loc[lower center] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_from_glyph - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_add_legend_loc[upper center] - AssertionError: Not all objects GCed:
ERROR tests/plotting/test_renderer.py::test_legend_from_multiple_glyph - AssertionError: Not all objects GCed:
================= 47 failed, 1391 passed, 10 skipped, 2 xfailed, 17 warnings, 41 errors in 346.34s (0:05:46) ================
```

</details>

### Pseudocode or Screenshots

Install the latest nightly build of VTK (locally I'm running off the `master` branch of VTK):

```bash
pip install vtk --pre --no-cache --extra-index-url https://wheels.vtk.org
```

**Comment by user27182** [comment]
`copy_from` is broken with vtk 9.4 when copying from an unwrapped object (e.g. `PolyData().copy_from(vtkPolyData())`).
I've tracked it down to an import issue. 

E.g. this change:

``` diff
- from vtkmodules.vtkCommonDataModel import vtkPolyData
+ from vtkmodules.util.data_model import vtkPolyData
```

resolves the `Text3DSource` failures: `TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.vtkPolyData'> must be compatible with the one being ... `

We might have to modify all related `vtkDataSet` imports

**Comment by user27182** [comment]
for this failure:
```
FAILED tests/core/test_grid.py::test_create_image_data_from_specs - AssertionError: assert ImageData (0x... Arrays:     0 == ImageData (0x... Arrays:     0
```
The cause is an issue with `ImageData.__eq__`. I.e. this assertion is failing:
https://github.com/pyvista/pyvista/blob/e7db352ce3f35eb12b13cfcc4b58a8123d7a8dc4/tests/core/test_grid.py#L858-L860

I have tracked this down to an issue with the `cells` attribute which is causing Line 361 here to evaluate to `True` and therefore return `False` (i.e. not equal):
https://github.com/pyvista/pyvista/blob/e7db352ce3f35eb12b13cfcc4b58a8123d7a8dc4/pyvista/core/dataobject.py#L351-L362
 With vtk<9.4, `ImageData` does not have a `cells` attribute:
``` python
hasattr(pv.ImageData(), 'cells')  # False
```
But with vtk9.4, this is no longer the case:
``` python
pv.ImageData(dimensions=(2,2,2)).cells
Out: <vtkmodules.vtkCommonDataModel.vtkStructuredCellArray(0x60000336e8e0) at 0x116d32860>
```

So, we will need to implement `ImageData.cells` so that it returns a NumPy array, not a vtkStructuredCellArray

I suspect this is also the root cause of the `test_pickle_multiblock` tests since those tests use equality to verify that the pickling is working (and any ImageData blocks in in the multiblock will return False)

**Comment by awoimbee** [comment]
v9.4 is relased, breaking pyvista
(personally getting `AttributeError: 'vtkmodules.vtkCommonCore.vtkPoints' object has no attribute 'dtype'`)

**Comment by banesullivan** [comment]
@awoimbee, any help or contributions you may be able to provide is welcome and appreciated!

**Comment by alechouse97** [comment]
Sure, here's some snippets:

```bash
$ uv pip show pyvista vtk
Using Python 3.12.7 environment at /home/alec.house/local/mambaforge/envs/sage
Name: pyvista
Version: 0.43.10
Location: /home/alec.house/local/mambaforge/envs/sage/lib/python3.12/site-packages
Requires: matplotlib, numpy, pillow, pooch, scooby, vtk
Required-by: sage
---
Name: vtk
Version: 9.4.0
Location: /home/alec.house/local/mambaforge/envs/sage/lib/python3.12/site-packages
Requires: matplotlib
Required-by: pyvista
```

```python
>>> import pyvista
>>> grid = pyvista.read("rae2822_template.vtu")
>>> type(grid)
<class 'pyvista.core.pointset.UnstructuredGrid'>
>>> type(grid.points)
<class 'vtkmodules.vtkCommonCore.vtkPoints'>

>>> grid.cell_centers()
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    grid.cell_centers()
  File "/home/alec.house/local/mambaforge/envs/sage/lib/python3.12/site-packages
/pyvista/core/filters/data_set.py", line 2146, in cell_centers
    return _get_output(alg)
           ^^^^^^^^^^^^^^^^
  File "/home/alec.house/local/mambaforge/envs/sage/lib/python3.12/site-packages
/pyvista/core/filters/__init__.py", line 48, in _get_output
    data.copy_meta_from(ido, deep=True)
  File "/home/alec.house/local/mambaforge/envs/sage/lib/python3.12/site-packages
/pyvista/core/dataset.py", line 1471, in copy_meta_from
    self._association_complex_names = deepcopy(ido._association_complex_names)
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alec.house/local/mambaforge/envs/sage/lib/python3.12/site-packages
/pyvista/core/dataset.py", line 150, in __getattr__
    return super().__getattribute__(item)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'UnstructuredGrid' object has no attribute '_association_complex_names'
```
Here's the .vtu file that is used in the above snippet: [rae2822_template.zip](https://github.com/user-attachments/files/17906765/rae2822_template.zip)

Hope this helps!

**Comment by banesullivan** [comment]
Ah okay, thanks for checking the upgrade! We had some preliminary work fixing our method resolution orders for VTK 9.4 compatibility that was released in v0.44.0 which is what you were seeing. Ref #5650

**Comment by ConnorNoddin** [comment]
```
TypeError: The Input DataSet type <class 'vtkmodules.util.data_model.PolyData'> must be compatible with the one being overwritten <class 'pyvista.core.pointset.PolyData'>
```

Issue when trying to create a pyvista.Text3D() object

**Comment by banesullivan** [comment]
For the meantime, we are going to limit VTK's version when installing PyVista via #6872

Hopefully we can resolve the issues from this VTK release prior to PyVista's next minor release (planned for December)

**Comment by banesullivan** [comment]
Reported an upstream issue with `vtkSampleFunction` which is causing our failing tests in `tests/core/test_imaging.py::test_sample_function` See https://gitlab.kitware.com/vtk/vtk/-/issues/19550

**Comment by banesullivan** [comment]
@Conor-PhysicsX, thank you for being willing to help and contribute here!! We don't have a concrete work plan at the moment but are still in a bit of reactionary problem solving mode. For now, I'm going through the tests one by one and seeing what the failures is a result of. If you have the patience to help with this that would be so helpful!! 

Right now, I'm trying to determine if each tests' failure is the result of

1. A different return type due to VTK's new Python data model and wrapping mechanism that exposes VTK methods as snake-case attributes on the data types. These are issues we will have to work around ourselves and is mostly what #6879 is about
2. A genuine regression in VTK (like the comment just above this one). If we find these, then we should open an issue upstream in VTK with a minimally reproducible example in pure VTK Python code.

Any insights you make, please post on this issue so that we can track these changes and determine how to go about addressing each. Again thank you for any help you are able to provide!!

**Comment by banesullivan** [comment]
I'm not totally confident, but it looks like this change https://gitlab.kitware.com/vtk/vtk/-/merge_requests/8265/diffs may have affected the test we have for flipping normals of a mesh in `tests/core/test_polydata.py::test_flip_normals` @rupertnash would be able to help us understand the changes made to `vtkPolyDataNormals` and if we're making assumptions in our `flip_normals` code or `compute_normals` code? Otherwise there have been a few other changes to `vtkPolyDataNormals` recently that may have caused this. https://gitlab.kitware.com/vtk/vtk/-/merge_requests?scope=all&state=merged&search=vtkPolyDataNormals I can run a git-bisect on VTK later if the issue isn't obvious.

https://github.com/pyvista/pyvista/blob/967d23a5f05647281608c5210d00a4007725991f/pyvista/core/filters/poly_data.py#L2620-L2642

https://github.com/pyvista/pyvista/blob/967d23a5f05647281608c5210d00a4007725991f/pyvista/core/filters/poly_data.py#L1586-L1753

**Comment by banesullivan** [comment]
It looks like there are a number of testing failures with `pad_image` and specifically validating the right output scalars were set as active. I'm struggling to track this down, @user27182 as the original author of `pad_image`, is this something you  might have availability to look into?

```
FAILED tests/core/test_imagedata_filters.py::test_pad_image_pad_size_bounds[pad_size0-expected_dimensions0-expected_bounds0] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_pad_size_bounds[pad_size1-expected_dimensions1-expected_bounds1] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_pad_size_bounds[pad_size2-expected_dimensions2-expected_bounds2] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_pad_size_bounds[pad_size3-expected_dimensions3-expected_bounds3] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_pad_size_bounds[pad_size4-expected_dimensions4-expected_bounds4] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_pad_size_bounds[pad_size5-expected_dimensions5-expected_bounds5] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_pad_size_bounds[pad_size6-expected_dimensions6-expected_bounds6] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_pad_size_bounds[pad_size7-expected_dimensions7-expected_bounds7] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_pad_size_bounds[pad_size8-expected_dimensions8-expected_bounds8] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_pad_size_bounds[pad_size9-expected_dimensions9-expected_bounds9] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_scalars[None-image-True] - AssertionError: assert 'other' == 'image'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_scalars[other-other-True] - AssertionError: assert 'image' == 'other'
FAILED tests/core/test_imagedata_filters.py::test_pad_image_multi_component - AssertionError: assert False
```

**Comment by user27182** [comment]
> is this something you might have availability to look into?

I can look into this. How can I reproduce these failures?

**Comment by banesullivan** [comment]
@user27182, to reproduce all you need to do is install VTK 9.4 from PyPI and run `pytest -v tests/core/test_imagedata_filters.py`

**Comment by user27182** [comment]
> @user27182, to reproduce all you need to do is install VTK 9.4 from PyPI and run `pytest -v tests/core/test_imagedata_filters.py`

I am unable to reproduce these failures with this. All `pad_image` tests are passing for me locally. I'm on macOS and using pyvista's `main` branch (and vtk 9.4.0). Is there a way to reproduce this with CI?

**Comment by banesullivan** [comment]
Oh! well how about that! They are indeed passing on main but failing on my branch for #6879. That may give us a clue, but I think I'm done investigating for the evening

**Comment by banesullivan** [comment]
So it looks like we are not be able to define any default arguments in the `__init__`'s of our wrapped VTK classes: https://gitlab.kitware.com/vtk/vtk/-/issues/19550#note_1598882

**Comment by banesullivan** [comment]
@user27182, I've pushed changes to my branch that disable the override mechanism and the issues with ImageData and `tests/core/test_imagedata_filters.py` have disappeared. See above link to the VTK issue.

**Comment by banesullivan** [comment]
All core API tests are currently passing for me on VTK 9.4 in #6879! 🎉  still a few failures for the plotting tests though

**Comment by cquammen** [comment]
> All core API tests are currently passing for me on VTK 9.4 in #6879! 🎉 still a few failures for the plotting tests though

Hi @banesullivan where can we be of service to help push this through?

**Comment by rupertnash** [comment]
Hi @banesullivan - I found a bug in vtkPolyDataNormals when AutoOrientNormals was on (https://gitlab.kitware.com/vtk/vtk/-/issues/18270) and (believe) I fixed it (https://gitlab.kitware.com/vtk/vtk/-/merge_requests/8265). As you can see from the code you linked, the changes were only to `vtkOrientPolyData`.

If you have a vtkPolyData that this fails for (i.e. AutoOrientNormals = true and the resulting normals point inwards), please add a bug to Kitware's issue tracker with the details and tag me in!

**Comment by banesullivan** [comment]
> Do you expect to be able to use PyVista with VTK 9.4.0? Or will https://gitlab.kitware.com/vtk/vtk/-/merge_requests/11698 in a future VTK 9.4.1 release be needed?

I'm expecting 9.4.0 to be compatible with PyVista once #6879 finishes. We're going to completely opt-out of the override mechanisms in VTK for now so that specific Merge Request in VTK isn't immediately necessary for us but will be required if we use the override mechanism with PyVista moving forward.

> where can we be of service to help push this through?

Thanks for offering to help push this through @cquammen!! I think we are very close in #6879, I just need to pick up where I left off (end of year is a busy time!) and outline what remains. I'll take some time this weekend to track down the remaining testing failures and deliver a clear list of what remains to be addressed.

**Comment by cquammen** [comment]
> We're going to completely opt-out of the override mechanisms in VTK for now so that specific Merge Request in VTK isn't immediately necessary for us but will be required if we use the override mechanism with PyVista moving forward.

Great, that's helpful information, @banesullivan.

> Thanks for offering to help push this through @cquammen!! I think we are very close in https://github.com/pyvista/pyvista/pull/6879, I just need to pick up where I left off (end of year is a busy time!) and outline what remains.

Understood. Thanks for the update!

**Comment by banesullivan** [comment]
Plotting tests that are still failing against VTK 9.4. Just need to go one by one and resolve these

```py
FAILED tests/plotting/test_charts.py::test_chart_2d - assert not True
FAILED tests/plotting/test_charts.py::test_axis_range - AssertionError: assert 0 == 'auto'
FAILED tests/plotting/test_charts.py::test_axis_scale - assert not True
FAILED tests/plotting/test_charts.py::test_axis_visible - assert not True
FAILED tests/plotting/test_charts.py::test_axis_tick_locations - TypeError: 'vtkmodules.vtkCommonCore.vtkStringArray' object is not iterable
FAILED tests/plotting/test_plotting.py::test_add_box_axes - pytest_pyvista.pytest_pyvista.RegressionError: test_add_box_axes Exceeded image regression error of 500.0 with an image error equal to: 1773.9568627450699
FAILED tests/plotting/test_plotting.py::test_add_axes_twice - pytest_pyvista.pytest_pyvista.RegressionError: test_add_axes_twice Exceeded image regression error of 500.0 with an image error equal to: 2543.9908496732915
FAILED tests/plotting/test_plotting.py::test_hide_axes - pytest_pyvista.pytest_pyvista.RegressionError: test_hide_axes Exceeded image regression error of 500.0 with an image error equal to: 2608.3725490197035
FAILED tests/plotting/test_plotting.py::test_add_axes_parameters - pytest_pyvista.pytest_pyvista.RegressionError: test_add_axes_parameters Exceeded image regression error of 500.0 with an image error equal to: 2487.057516339956
FAILED tests/plotting/test_plotting.py::test_plot_composite_poly_scalars_opacity - pytest_pyvista.pytest_pyvista.RegressionError: test_plot_composite_poly_scalars_opacity Exceeded image regression error of 500.0 with an image error equal to: 948.25490...
FAILED tests/plotting/test_plotting.py::test_plot_composite_poly_scalars_cell - pytest_pyvista.pytest_pyvista.RegressionError: test_plot_composite_poly_scalars_cell Exceeded image regression error of 500.0 with an image error equal to: 2455.8052287...
FAILED tests/plotting/test_plotting.py::test_legend_scale - DeprecationWarning: Call to deprecated method SetLabelModeToXYCoordinates. (This class can now determine current plane. Please use the generic SetLabelModeToCoordinates...
```

**Comment by banesullivan** [comment]
Main issue we are seeing appears to be some sort of weird bug in VTK when removing and re-adding the axes widget that causes a blue square to appear in the lower left corner:

![Image](https://github.com/user-attachments/assets/c0329422-7e2d-4026-84f9-9f08cfdfc9a1)

I've managed to reproduce in pure Python VTK as:

```py
import vtk

ren = vtk.vtkRenderer()
renWin = vtk.vtkRenderWindow()
renWin.AddRenderer(ren)
iren = vtk.vtkRenderWindowInteractor()
style = vtk.vtkInteractorStyleTrackballCamera()
iren.SetRenderWindow(renWin)
iren.SetInteractorStyle(style)

axes = vtk.vtkAxesActor()
widget = vtk.vtkOrientationMarkerWidget()
widget.SetOrientationMarker(axes)
widget.SetInteractor(iren)
widget.SetViewport(0.0, 0.0, 0.2, 0.2)
widget.SetEnabled(1)
widget.InteractiveOn()

widget.EnabledOff()
del widget

widget = vtk.vtkOrientationMarkerWidget()
widget.SetOrientationMarker(axes)
widget.SetInteractor(iren)
widget.SetViewport(0.0, 0.0, 0.2, 0.2)
widget.SetEnabled(1)
widget.InteractiveOn()

iren.Initialize()
renWin.Render()
iren.Start()
del iren, renWin
```

If I modify this to reset the widgets viewport to `widget.SetViewport(0.0, 0.0, 0.0, 0.0)` right before calling `widget.EnabledOff()` then the blue box disappears, but I get some OpenGL errors: `vtkOpenGLTexture.cxx:195    ERR| vtkOpenGLTexture (0x129713f30): No scalar values found for texture input!`

Setting `widget.SetViewport(0.0, 0.0, 0.01, 0.01)` seems to work sufficiently well here though. I'm going to do this whenever we attempt to remove the `vtkOrientationMarkerWidget` so as to avoid this.... total hack. @cquammen, perhaps you could report this issue upstream in VTK?

**Comment by jspanchu** [comment]
Fix is in https://gitlab.kitware.com/vtk/vtk/-/merge_requests/11874

**Comment by MatthewFlamm** [comment]
Latest run against vtk master had only this failed test 

> FAILED tests/plotting/test_plotting.py::test_box_axes - RuntimeError: test_box_axes Exceeded image regression error of 500.0 with an image error equal to: 944.5398692810609
= 1 failed, 2237 passed, 16 skipped, 2 xfailed, 10 warnings, 1 error in 1160.17s (0:19:20) =

**Comment by user27182** [comment]
> Latest run against vtk master had only this failed test
> 
> > FAILED tests/plotting/test_plotting.py::test_box_axes - RuntimeError: test_box_axes Exceeded image regression error of 500.0 with an image error equal to: 944.5398692810609
> > = 1 failed, 2237 passed, 16 skipped, 2 xfailed, 10 warnings, 1 error in 1160.17s (0:19:20) =

Are you referring to this workflow?
https://github.com/pyvista/pyvista/actions/runs/13427743158/job/37513756651#step:9:2706
If you look at the software report, this is only for vtk 9.3.1. I could be wrong but I'm not seeing any tests against VTK Master for vtk 9.4.

**Comment by johnnynunez** [comment]
vtk arm is merged on vtk

**Comment by Oblynx** [comment]
[VTK 9.4.2 has just been released!](https://gitlab.kitware.com/vtk/vtk/-/tags/v9.4.2)

**Comment by ConnorNoddin** [comment]
Is this closed now? And, when is the next planned Pyvista release?

Thanks! <3

**Comment by johnnynunez** [comment]
VTK 9.5 aarch64 wheels! https://vtk.org/files/release/9.5/ @banesullivan

**Comment by xmy601** [comment]
I encountered an error：AttributeError: 'UnstructuredGrid' object has no attribute '_association_complex_names'
my python is 3.9, and the vtk is 9.5.1, the pyvista version is 0.46.3, how should I modify it?

**Comment by MatthewFlamm** [comment]
> I encountered an error：AttributeError: 'UnstructuredGrid' object has no attribute '_association_complex_names' my python is 3.9, and the vtk is 9.5.1, the pyvista version is 0.46.3, how should I modify it?

This feels like a pure PyVista bug.  Can you please open a new issue with the full traceback?

**Comment by xmy601** [comment]
i have open a new issue&nbsp;




徐明月
***@***.***



        



         原始邮件
         
       
发件人：MatthewFlamm ***@***.***&gt;
发件时间：2025年9月16日 19:01
收件人：pyvista/pyvista ***@***.***&gt;
抄送：xmy601 ***@***.***&gt;, Comment ***@***.***&gt;
主题：Re: [pyvista/pyvista] VTK 9.4 Incompatibility Tracker (Issue #6731)



MatthewFlamm left a comment (pyvista/pyvista#6731)

I encountered an error：AttributeError: 'UnstructuredGrid' object has no attribute '_association_complex_names' my python is 3.9, and the vtk is 9.5.1, the pyvista version is 0.46.3, how should I modify it?

This feels like a pure PyVista bug.  Can you please open a new issue with the full traceback?
